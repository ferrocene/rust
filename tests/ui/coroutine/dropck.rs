#![feature(coroutines, coroutine_trait, stmt_expr_attributes)]

use std::cell::RefCell;
use std::ops::Coroutine;
use std::pin::Pin;

fn main() {
    let (mut coro, cell);
    cell = Box::new(RefCell::new(0));
    let ref_ = Box::leak(Box::new(Some(cell.borrow_mut())));
    //~^ ERROR `*cell` does not live long enough [E0597]
    // the upvar is the non-dropck `&mut Option<Ref<'a, i32>>`.
    coro = #[coroutine]
    || {
        // but the coroutine can use it to drop a `Ref<'a, i32>`.
        let _d = ref_.take();
        yield;
    };
    Pin::new(&mut coro).resume(());
    // drops the RefCell and then the Ref, leading to use-after-free
}
