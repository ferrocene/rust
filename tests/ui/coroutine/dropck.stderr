error[E0597]: `*cell` does not live long enough
  --> $DIR/dropck.rs:10:40
   |
LL |     let (mut coro, cell);
   |                    ---- binding `cell` declared here
LL |     cell = Box::new(RefCell::new(0));
LL |     let ref_ = Box::leak(Box::new(Some(cell.borrow_mut())));
   |                                        ^^^^ borrowed value does not live long enough
...
LL | }
   | -
   | |
   | `*cell` dropped here while still borrowed
   | borrow might be used here, when `coro` is dropped and runs the destructor for coroutine
   |
   = note: values in a scope are dropped in the opposite order they are defined

error: aborting due to 1 previous error

For more information about this error, try `rustc --explain E0597`.
